File: 572-subtree-of-another-tree/subtree-of-another-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if(subRoot == null)
        {
            return true;
        }
        if(root==null)
        {
            return false;
        }
        if(present(root,subRoot))
        {
            return true;
        }
        return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
    }
    private boolean present(TreeNode a, TreeNode b){
        if(a == null && b ==null) 
        {
            return true;
        }
        if(a == null || b == null) 
        {
            return false;
        }
        if(a.val != b.val) 
        {
            return false;
        }
        return present(a.left, b.left) && present(a.right, b.right);
    }
}

File: 572-subtree-of-another-tree/subtree-of-another-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        
        def isIdentical(n1, n2):
            if not n1 and not n2: return True
            if not n1 or not n2 or n1.val != n2.val: return False
            return isIdentical(n1.left, n2.left) and isIdentical(n1.right, n2.right)

        if not subRoot: return True
        if not root: return False
        
        if isIdentical(root, subRoot): return True
        
        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)